webapp/javascript/components/TimelineChart/Annotations.plugin.tsx

```diff
@@ -0,0 +1,285 @@
+    return x >= an.fromX && x <= an.toX && y >= an.fromY && y <= an.toY;
+  });
+};
+
+function drawRoundRect(
+  ctx: CtxType,
+  x: number,
+  y: number,
+  w: number,
+  h: number,
+  radius: number,
+  color: string | Color
+) {
+  const r = x + w;
+  const b = y + h;
+  ctx.beginPath();
+  ctx.strokeStyle = color;
+  ctx.fillStyle = color;
+  ctx.lineWidth = 1;
+  ctx.moveTo(x + radius, y);
+  ctx.lineTo(r - radius, y);
+  ctx.quadraticCurveTo(r, y, r, y + radius);
+  ctx.lineTo(r, y + h - radius);
+  ctx.quadraticCurveTo(r, b, r - radius, b);
+  ctx.lineTo(x + radius, b);
+  ctx.quadraticCurveTo(x, b, x, b - radius);
+  ctx.lineTo(x, y + radius);
+  ctx.quadraticCurveTo(x, y, x + radius, y);
+  ctx.fill();
+  ctx.stroke();
+}
+
+(function ($) {
+  function init(plot: jquery.flot.plot & jquery.flot.plotOptions) {
+    const annotationsPositions: AnnotationPosition[] = [];
+    const placeholder = plot.getPlaceholder();
+
+    function onHover(_: unknown, pos: { x: number; y: number }) {
+      if (annotationsPositions?.length) {
+        const { x, y } = getCursorPositionInPx(plot, pos);
+
+        const annotation = findAnnotationByCursorPosition(
+          x,
+          y,
+          annotationsPositions
+        );
+
+        if (annotation) {
+          placeholder.trigger('hoveringOnAnnotation', [{ hovering: true }]);
+        } else {
+          placeholder.trigger('hoveringOnAnnotation', [{ hovering: false }]);
+        }
+      }
+    }
+
+    function onClick(
+      _: unknown,
+      pos: { x: number; pageX: number; pageY: number; y: number }
+    ) {
+      const options: IFlotOptions = plot.getOptions();
+      const container = inject($);
+      const containerEl = container?.[0];
+
+      ReactDOM.unmountComponentAtNode(containerEl);
+
+      const ContextMenu = options?.ContextMenu;
+
+      const { x, y } = getCursorPositionInPx(plot, pos);
+
+      const annotation = findAnnotationByCursorPosition(
+        x,
+        y,
+        annotationsPositions
+      );
+
+      if (ContextMenu && containerEl) {
+        const timestamp = Math.round(pos.x / 1000);
+        const value = annotation
+          ? {
+              timestamp: annotation.timestamp,
+              content: annotation.content,
+            }
+          : null;
+
+        ReactDOM.render(
+          <ContextMenu
+            click={{ ...pos }}
+            containerEl={containerEl}
+            timestamp={timestamp}
+            value={value}
+          />,
+          containerEl
+        );
+      }
+    }
+
+    plot.hooks!.draw!.push((plot, ctx: CtxType) => {
+      const o: IFlotOptions = plot.getOptions();
+
+      if (o.annotations?.length) {
+        const axes = plot.getAxes();
+        const plotOffset: { top: number; left: number } = plot.getPlotOffset();
+        const extractedX = extractRange(plot, axes, 'x');
+        const extractedY = extractRange(plot, axes, 'y');
+
+        o.annotations.forEach((a: AnnotationType) => {
```

Comment:
As I said in some other comment, when reading the functionality one doesn't really want to be concerned about canvas drawing. Move this to its own function please.